lib/rspec/support/fuzzy_matcher.rb

```diff
@@ -6,9 +6,11 @@ module Support
     module FuzzyMatcher
       # @api private
       def self.values_match?(expected, actual)
-        if Hash === actual
-          return hashes_match?(expected, actual) if Hash === expected
-        elsif Array === expected && Enumerable === actual && !(Struct === actual)
+        if actual.instance_of?(Hash)
+          return hashes_match?(expected, actual) if expected.instance_of?(Hash)
+        elsif expected.instance_of?(Array) && actual.instance_of?(Array)
+          return arrays_match?(expected, actual)
+        elsif expected.instance_of?(Array) && Enumerable === actual && !(Struct === actual)
```

Comment:
OK. The problem with `Class#===` is that it returns `true` not just for instances of `klass`, but for instances of *subclasses* of `klass`. The whole point of this PR is that `FuzzyMatcher`'s matching doesn't work well with subclasses of `Array` which may have redefined `#to_a` or `#==`.

This is a tricky situation! We're working in a very dynamic language, where any method on any object can be redefined to do anything at all. I see why you like `Class#===`. Because it is defined on `Class`, and not the object being tested, you don't need to make any assumptions at all about what crazy things might have been done to the object under test.

An alternative which works in Ruby 2.0 and up is `Kernel.instance_method(:instance_of?).bind(obj).call(Array)`.

Another alternative is to use an interpreter-specific extension module which reaches "under the hood" to pick out the actual class of an object. With this technique, you can write code which works even with `BasicObject` instances, or other classes which have had the built-in `Object` methods redefined to do crazy things.

Any thoughts?